Add nixpkgs.config and overlays option

Gerschtli opened the pull request:
Now it is possible to define config and overlays in nix-on-droid config which is a necessary step for https://github.com/t184256/nix-on-droid-bootstrap/issues/23

t184256 commented on modules/nixpkgs.nix:
```diff
@@ -0,0 +1,137 @@
+# Licensed under GNU Lesser General Public License v3 or later, see COPYING.
+# Copyright (c) 2019 Alexander Sosedkin and other contributors, see AUTHORS.
+
+# Adapted from nixpkgs/home-manager.
```
Given it's MIT-licensed, you might have to embed the original license header as described in
http://softwarefreedom.org/resources/2007/gpl-non-gpl-collaboration.html,
section 2.2 "Adding GPL'd modifications to permissive-licensed files".

Gerschtli replied:
Is there a possibility to add just a small one-/two-liner like your gpl header and add the complete license add the repo root? Because I would rather not paste 20 or more useless comment lines as this should be done for some more files already present in nix-on-droid.. But I never really got in touch with these license things.

t184256 commented on modules/nixpkgs.nix:
```diff
@@ -0,0 +1,137 @@
+          </para><para>
+
+          Note, this option will not apply outside your Nix-on-Droid
+          configuration like when installing manually through
```
And it won't apply to home-manager configuration either. Should it? If not, should we probably state that too and provide a snippet to pass it through to it?

Same goes for overlays.

Gerschtli replied:
Like in nixpkgs there is also no such comment, but I can add a simple note

Gerschtli commented:
I just added two new options to inject config and overlays in home-manager config

t184256 commented:
Uhh, does this warrant separate options? I was hoping for some single-line trick in the doctext, this seems to be an overkill.

Gerschtli commented:
You mean one flag for inject config and overlays instead of two?

t184256 commented:
No, I meant something like (warning: I'm totally making this up on the spot)

```
let
  overlays = ...
in
  {
   nixpkgs.overlays = overlays;
   home-manager.config.nixpkgs.overlays = overlays;
  }
```
```
let
  nixpkgs.overlays = ...
in
  {
    inherit nixpkgs.overlays;
    home-manager.config = {
      inherit nixpkgs.overlays;
    };
  }
```

or even just

```
rec {
  nixpkgs.overlays = ...
  nixpkgs.config = ...
  home-manager.config = {
    inherit nixpkgs;
 };
}
```

if anything of this makes any sense.

Gerschtli commented:
The Problem with this is, you can not use home-manager.config as attribute set, it is a home-manager module and therefore a function **or** an attribute set.. What we could do (or to be more clear: try) is to write custom option type for the hm config which injects those attributes into the config function, but I don't know how we could achieve this..

Or add an option like home-manager.configExtra which is an attribute set which we could inject into the home-manager config.

There are plenty of ways, the question is, what is the goal? Do we want a generic way to set arbitrary options in hm config or is nixpkgs.config/.overlays enough?

What are the up/downsides of your approach compared to mine?

t184256 commented:
The Problem with this is, you can not use home-manager.config as attribute set, it is a home-manager module and therefore a function or an attribute set..

Sorry, I don't get it. I'm not trying to use it as an attribute set, I'm trying to inject values into it with `inherit`.

```
nix-repl> ((let nested.attr = "value"; in rec { func = param: {inherit nested;}; } ).func "arg").nested.attr
"value"
```

> There are plenty of ways, the question is, what is the goal?

I wanted to just add a hint to the doc: 'to set these values inside h-m config, you can access the outer ones in this manner'.

> What are the up/downsides of your approach compared to mine?

Pros:

* flexible
* clear
* no options required, you just do what you meant to do: set this to that using existing Nix constructs

Cons:

* no idea if this is possible
* I clearly cannot communicate my idea

Will try it later and post a working config snippet if I manage to produce it.

t184256 commented:
Passing the fixed point of the outer config works:
```
{pkgs, config, ...}:
{
  nixpkgs.overlays = [ (self: super: { unstable = import <nixos-unstable> {}; })];
  home-manager.config = {pkgs, ...}:
  {
    inherit (config) nixpkgs;
  };
}
```
The only downside I see is potential shadowing by the `config` fixed point of the h-m config. Will think more.

t184256 commented:
```
{pkgs, config, ...}:
{
  nixpkgs.overlays = [ (self: super: { unstable = import <nixos-unstable> {}; })];
  home-manager.config = let outer-config = config; in {pkgs, config ...}:
  {
    # now both outer-config and config are available, one can do, e.g.,
    inherit (outer-config) nixpkgs;
  };
}
```
Anything wrong with that one?

Gerschtli commented:
Yes I agree, this works, but only if you define the `home-manager.config` inline. In my setup (https://github.com/Gerschtli/dotfiles/blob/master/nix-on-droid/nix-on-droid.nix) I import my `home.nix` file. There I don't have access to the outer config. We could add something like `nix-on-droid-config` as parameter for each `home-manager` module via `_module.args`.

Gerschtli commented:
Somehow this works:

```nix
{ pkgs, lib, config, ...}:

lib.mkMerge [
  {
    home-manager.config.nixpkgs.config = import ./nixpkgs-config.nix;
    nixpkgs.config = import ./nixpkgs-config.nix;
  }
  {
    home-manager.config =
      { pkgs, ... }:
      {
        # other config
      };
  }
]
```

So I removed the inject options and updated the option descriptions.

t184256 commented:
Isn't this just

```
{ pkgs, lib, config, ...}:

  {
    home-manager.config.nixpkgs.config = import ./nixpkgs-config.nix;
    nixpkgs.config = import ./nixpkgs-config.nix;
    home-manager.config =
      { pkgs, ... }:
      {
        # other config
      };
  }
```

?

Gerschtli commented:
You cannot set an option twice AFAIK. Therefore the mkMerge in my example.

t184256 commented:
Automatically merging values from multiple assignments into attribute sets is perfectly, but now I see how you're merging an attrset with a function-returning-an-attrset, and I have no idea what's going on here.

Gerschtli commented:
This is the part where I have no clue what is happening, but it works.. :D I'm open for better ideas but I think for now we can leave it like this as we are not defining any new options.

t184256 commented:
Yeah, you are right. My objection was adding extraneous options, and now it's fine to land this without figuring out the perfect docstring from the first attempt.
